### PR TITLE
LibraryManager yield, and reduced FileMonitors

### LINKED ISSUE #139: Videos takes time to launch on large, non-default library-folder
## Prerequisites
- [x] I have searched open and closed issues for duplicates.

## Describe the bug
Videos takes time to launch on **large, non-default** library-folder. Users might delete the default library folder and symlink it to home directory. Relevant Discussion (https://github.com/elementary/videos/issues/137). Takes upto 22 minutes in some cases.

### To Reproduce
1. Set library-folder to /home/<username> `gsettings set io.elementary.videos library-folder "/home/<username>"`
2. Launch Videos

### Expected behavior
Videos is launched immediately.

## Platform Information
OS: Juno

- [x] I'm using the latest version from git that I've manually compiled
- [ ] I'm using the latest released stable version

## Additional context
Can be reproduced by pointing library folder to Steam Library.

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/81767888-videos-takes-time-to-launch-on-large-non-default-library-folder?utm_campaign=plugin&utm_content=tracker%2F61134914&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F61134914&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by j60tt0**
Fixes #139 
In order to address the slow start-up and extremely high CPU usage associated with https://github.com/elementary/videos/issues/139 I have made a few changes to the LibraryManager which allow it to yield back to the main thread to allow faster start-up, yield before each iteration to allow better interleving with the playback, and a reduction of the FileMonitor usage to just directories containing Videos. These changes should allow better background scanning of the Library without requiring explicit worker threads, and ensure that people who have a non-standard library location with a large number of directories to traverse, do so without preventing the normal operation of the application,

### CODE REVIEW

**Comment by jeremypw** [comment]
I can confirm it speeds up start up under the circumstances described in the issue report.

File: src/Services/LibraryManager.vala [hunk 1]
```diff
@@ -67,57 +70,85 @@ namespace Audience.Services {
...
+                    return;
+                }
+                if (file_info.get_file_type () == FileType.DIRECTORY) {
+                    unchecked_directories.offer (src.get_path ());
+                    if (!is_scanning) {
+                        begin_scan ();
                     }
+                } else if (is_file_valid (file_info)) {
+                    string src_path = src.get_path ();
+                    crate_video_object (file_info, Path.get_dirname (src_path), Path.get_basename (src_path));
```
Comment (jeremypw) [inline src/Services/LibraryManager.vala:101]: Spelling - `create_video_object`?

Comment (j60tt0) [reply src/Services/LibraryManager.vala]: Yes, it seems that was the intended name: renamed.

File: src/Services/LibraryManager.vala [hunk 2]
```diff
@@ -67,65 +70,93 @@ namespace Audience.Services {
...
+                try {
+                    file_info = src.query_info (FileAttribute.STANDARD_CONTENT_TYPE + "," + FileAttribute.STANDARD_IS_HIDDEN + "," + FileAttribute.STANDARD_TYPE, 0);
+                } catch (Error e) {
+                    warning (e.message);
+                    return;
+                }
+                if (file_info.get_file_type () == FileType.DIRECTORY) {
+                    unchecked_directories.offer (src.get_path ());
+                    if (!is_scanning) {
+                        begin_scan ();
```
[the remark this answers is not in the document]
Comment (j60tt0) [reply src/Services/LibraryManager.vala]: Excellent point, that should have been a call to detect_video_files.begin ()

**Pull request merged**